lib/ansible/modules/packaging/os/package_facts.py

```diff
@@ -140,94 +145,231 @@
         }
 '''
 
+import json
 import sys
 
 from ansible.module_utils.basic import AnsibleModule
-from ansible.module_utils._text import to_text
-
-
-def rpm_package_list():
-
-    try:
-        import rpm
-    except ImportError:
-        module.fail_json(msg='Unable to use the rpm python bindings, please ensure they are installed under the python the module runs under')
-
-    trans_set = rpm.TransactionSet()
-    installed_packages = {}
-    for package in trans_set.dbMatch():
-        package_details = dict(name=package[rpm.RPMTAG_NAME],
-                               version=package[rpm.RPMTAG_VERSION],
-                               release=package[rpm.RPMTAG_RELEASE],
-                               epoch=package[rpm.RPMTAG_EPOCH],
-                               arch=package[rpm.RPMTAG_ARCH],
-                               source='rpm')
-        if package_details['name'] not in installed_packages:
-            installed_packages[package_details['name']] = [package_details]
-        else:
-            installed_packages[package_details['name']].append(package_details)
-    return installed_packages
-
-
-def apt_package_list():
-
-    try:
-        import apt
-    except ImportError:
-        module.fail_json(msg='Unable to use the apt python bindings, please ensure they are installed under the python the module runs under')
-
-    apt_cache = apt.Cache()
-    installed_packages = {}
-    apt_installed_packages = [pk for pk in apt_cache.keys() if apt_cache[pk].is_installed]
-    for package in apt_installed_packages:
-        ac_pkg = apt_cache[package].installed
-        package_details = dict(name=package, version=ac_pkg.version, arch=ac_pkg.architecture, source='apt')
-        if package_details['name'] not in installed_packages:
-            installed_packages[package_details['name']] = [package_details]
-        else:
-            installed_packages[package_details['name']].append(package_details)
-    return installed_packages
-
-
-# FIXME: add more listing methods
-def main():
-    global module
-    module = AnsibleModule(argument_spec=dict(manager=dict()), supports_check_mode=True)
-    manager = module.params['manager']
-    packages = {}
-    results = {}
+from ansible.module_utils._text import to_native, to_text
+
+
+class PkgMgr:
+
+    def __init__(self, module):
+
+        self.m = module
+
+    def test(self):
+        ''' This method is supposed to return True/False if the package manager is installed/usable by this module '''
+        raise NotImplementedError
```

Comment:
When you use ABCMeta and abstractmethod, you should just have these ```pass```.  That way an implementor can call the base class's method if they want and it won't raise an exception.

Reply:
i had reverted from that as per previous suggestions, any drawbacks of doing it with ABCMeta?